### PR TITLE
Add extra-metadata options to complete middleware

### LINKED ISSUE #176: :complete op to optionally return info for each var?
Hello,

I'm contemplating moving the Counterclockwise "server" code to cider-nrepl. Currently toying with it, but I already am able to use the :complete op instead of CCW's own code.

What's missing is that currently CCW returns not only a vector of var Strings, but a vector of metadata for each var, removing the need to do N+1 calls.

Granted, this is far more data to transport than just the current vector of Strings. I haven't done any benchmark yet.

Question is: would you find the idea of optionally returning a vector with the vars info acceptable?

Maybe you already know this is not feasible in practice because you already tried it or the function to return individual var metadata is too costly to call it ahead of time, etc. ?

### PR DESCRIPTION
**Pull request by Deraen**
Fixes #176 

Merged: alexander-yakushev/compliment#26
Merged: gtrak/cljs-tooling/pull/18
Merged: gtrak/cljs-tooling/pull/19

@bbatsov In relation to alexander-yakushev/compliment#26

I decided to open this PR here for comments even though PR to Compliment is not merged yet and cljs-tooling doesn't have the support for options yet. 

Added options which allow frontends to request arglists and doc for each completion results. Passes the options to Compliment.

### CODE REVIEW

**Comment by cichli** [comment]
This looks good to me, thanks for tackling the CLJS side of this as well!

File: src/cider/nrepl/middleware/complete.clj [hunk 1]
```diff
@@ -57,7 +60,8 @@
      :requires {"symbol" "The symbol to lookup"
                 "ns" "The symbol's namespace"
                 "session" "The current session"}
-     :optional {"context" "Completion context for compliment."}
+     :optional {"context" "Completion context for compliment."
+                "extra-metadata" "List of extra-metadata fields"}
```
Comment (cichli) [inline src/cider/nrepl/middleware/complete.clj]: Maybe we should document the possible values `extra-metadata` can contain here too. AFAICT it's just `:arglists` and `:doc` for now?

**Comment by Deraen** [comment]
Both PRs are merged.

@alexander-yakushev I think we need a new compliment release/snapshot?

**Comment by bbatsov** [comment]
@Deraen This could use a few specs.

**Comment by alexander-yakushev** [comment]
And by specs he means tests.
(Automatic bbatsov translator 0.1.0)

**Comment by Deraen** [comment]
Sure. I see there are tests for cljs complete I can extend but there seems to be no tests for clj complete yet? 

I'll update the PR in the evening.

**Comment by bbatsov** [comment]
> Sure. I see there are tests for cljs complete I can extend but there seems to be no tests for clj complete yet?

Hmm, thought we had some - if we don't guess you can use the cljs tests as a starting point.

**Comment by bbatsov** [comment]
Some tests are failing, although the errors are pretty confusing to me. I've restarted the build. Btw, our conventions mandate the use of present simple tense for commit message titles.

**Comment by Deraen** [comment]
Some tests are failing as cljs-tooling needs a fix to work with cljs-3196. And looks like some tests are failing because travis problems (Problem opening jar).

**Comment by alexander-yakushev** [comment]
Seen that one before, restarting the tests worked for me.

**Comment by expez** [comment]
I've restarted this build twice now, but it keeps failing with the same problems.

**Comment by bbatsov** [comment]
@cichli You're our resident ClojureScript expert. What's your advise regarding cljs-3196?

**Comment by cichli** [comment]
Sorry for the delayed reply, totally missed this notification... the Travis caching problems are gone now. As @Deraen says, we just need gtrak/cljs-tooling#19 to be resolved and a new cljs-tooling release to be cut to fix the failing tests.

**Comment by Deraen** [comment]
I reworded commit messages. Cljs-tooling change is hopefully now ok and is just waiting for @gtrak to check it.

**Comment by Deraen** [comment]
Updated with new cljs-tooling dep and everything should be working now.

**Comment by bbatsov** [comment]
:+1:

**Pull request merged**